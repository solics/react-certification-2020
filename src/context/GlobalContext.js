import React, { createContext, useReducer } from 'react'
import { globalReducer, globalInitialState } from '../reducers'

const GlobalContext = createContext()

function GlobalContextProvider({ children }) {
	const [state, dispatch] = useReducer(globalReducer, globalInitialState)
	return (
		<GlobalContext.Provider value={[state, dispatch]}>{children}</GlobalContext.Provider>
	)
}

export { GlobalContext, GlobalContextProvider }
