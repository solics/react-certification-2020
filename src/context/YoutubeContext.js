import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const [loading, setLoading] = useState(false)
	const [loadingRelated, setLoadingRelated] = useState(false)
	const [errorMsg, setErrorMsg] = useState('')
	const [errorMsgRelated, setErrorMsgRelated] = useState('')
	const [videosRelated, setVideosRelated] = useState([])

	const search = term => {
		setLoading(true)
		fetch(`${SEARCH_URL}q=${term}`)
			.then(response => response.json())
			.then(data => {
				setLoading(false)
				setErrorMsg('')
				if (data?.items?.length) {
					setVideos(data.items)
					saveArrayOnLocalStorage('videos', data.items)
				} else {
					setVideos([])
					setErrorMsg(data.error.message)
				}
			})
			.catch(error => {
				setLoading(false)
				setErrorMsg(error.error.message)
				console.error(error)
			})
	}

	const getRelatedVideos = videoId => {
		setLoadingRelated(true)
		fetch(`${SEARCH_URL}relatedToVideoId=${videoId}`)
			.then(response => response.json())
			.then(data => {
				setLoadingRelated(false)
				setErrorMsgRelated('')
				if (data?.items?.length) {
					setVideosRelated(data.items)
				} else {
					setVideosRelated([])
					setErrorMsgRelated(data.error.message)
				}
			})
			.catch(error => {
				setLoadingRelated(false)
				setErrorMsgRelated(error.error.message)
				console.error(error)
			})
	}

	const defaultContext = {
		videos,
		videosRelated,
		search,
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
		getRelatedVideos,
		loading,
		loadingRelated,
		errorMsg,
		errorMsgRelated,
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
