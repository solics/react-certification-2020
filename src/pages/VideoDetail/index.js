import React, { useContext, useEffect } from 'react'
import styled from 'styled-components/macro'
import { GlobalContext } from '../../context/GlobalContext'
import useVideoDetail from '../../customHooks/useVideoDetail'
import THEMES from '../../themes'
import RelatedVideos from '../../components/RelatedVideos'
import { addFavoriteVideo } from '../../utils/localStorage'

const VideoDetailStyled = styled.div`
	padding: 20px 80px;
	display: flex;
	flex-wrap: wrap;
`
const VideoInfoStyled = styled.div`
	padding-right: 40px;
	color: ${props => props.theme.color};
	width: 70%;
	min-width: 70%;
	@media only screen and (max-width: 768px) {
		width: 100%;
		padding-right: 0px;
	}
`

const IframeStyled = styled.iframe`
	width: 100%;
	min-height: 500px;
	border: 0;
	@media only screen and (max-width: 1100px) {
		min-height: 300px;
	}
`

export default function VideoDetail() {
	const [state] = useContext(GlobalContext)
	const [requestVideoDetail, videoId] = useVideoDetail()

	const {
		currentTheme,
		youtube: {
			currentVideo: { video, isLoading },
		},
	} = state

	useEffect(() => {
		requestVideoDetail()
	}, [requestVideoDetail, videoId])

	const {
		snippet: { title, channelTitle, description },
	} = video

	const addToFavorites = () => {
		addFavoriteVideo(video)
	}
	return (
		<VideoDetailStyled>
			{isLoading ? (
				<span>Loading</span>
			) : (
				<VideoInfoStyled theme={THEMES[currentTheme]}>
					<IframeStyled title={title} src={`https://www.youtube.com/embed/${videoId}`} />
					<h2>{title}</h2>
					<button type="button" onClick={addToFavorites}>
						Add to favorites
					</button>
					<h4>{channelTitle}</h4>
					<p>{description}</p>
				</VideoInfoStyled>
			)}
			<RelatedVideos theme={THEMES[currentTheme]} videoId={videoId} />
		</VideoDetailStyled>
	)
}
