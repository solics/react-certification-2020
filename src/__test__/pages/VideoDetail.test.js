import { render } from '@testing-library/react'
import { Route, Router } from 'react-router'
import { createMemoryHistory } from 'history'
import React from 'react'
import { GlobalContextProvider } from '../../context/GlobalContext'
import VideoDetail from '../../pages/VideoDetail'

const history = createMemoryHistory()
history.push('/video-detail/testId123456')

jest.mock('react-router', () => ({
	...jest.requireActual('react-router'),
	useParams: () => ({
		videoId: 'testId123456',
	}),
}))

export const renderWithRouter = Component =>
	render(
		<GlobalContextProvider>
			<Router history={history}>
				<Route>
					<Component />
				</Route>
			</Router>
		</GlobalContextProvider>
	)

describe('VideoDetail testing', () => {
	it('Video Detail Snapshot', () => {
		const { container } = renderWithRouter(VideoDetail)
		expect(container).toMatchSnapshot()
	})
})
