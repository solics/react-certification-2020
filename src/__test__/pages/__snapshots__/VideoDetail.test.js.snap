// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail testing Video Detail Snapshot 1`] = `
<div>
  <div
    class="VideoDetailStyled-sc-1yszey5-0 khFVQG"
  >
    <span>
      Loading
    </span>
    <div>
      loading...
    </div>
  </div>
</div>
`;
